lib/async_hooks.js

```diff
@@ -207,6 +208,128 @@ class AsyncResource {
   }
 }
 
+// AsyncContext part
+const idResourcePairing = new Map();
+const contexts = new Set();
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // save new resource
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResourceWeak = idResourcePairing.get(triggerAsyncId);
+    contexts.forEach((context) => {
+      context._propagate(resource, triggerResourceWeak && triggerResourceWeak.get(), triggerAsyncId);
+    });
+  },
+  destroy(asyncId) {
+    idResourcePairing.delete(asyncId);
+  },
+  promiseResolve(asyncId) {
+    idResourcePairing.delete(asyncId);
+  }
+});
+
+class AsyncContext {
+
+  constructor() {
+    // WeakMap resource -> store
+    this._resourceStore = new WeakMap();
+    this.count = 0;
+    this._firstCalls = new Map();
+  }
+
+  /**
+   * propagate the context from a parent resource to a child one
+   */
+  _propagate(resource, triggerResource, triggerAsyncId) {
```

Comment:
Can this propagation be rethought via prototypical inheritance instead (`const child = Object.create(parent)`)? This seems a lot of complicated code in a hot path.

Reply:
I will do some tests to check that!